RecoTauTag/HLTProducers/src/L2TauTagNNProducer.cc

```diff
@@ -0,0 +1,891 @@
+     l1t::TauVectorRef allTaus;
+
+     for (size_t inp_idx = 0; inp_idx<L1TauDesc.size(); inp_idx++){
+       CollectionNamesMap.push_back(L1TauDesc[inp_idx].CollectionName) ;
+
+       edm::Handle<trigger::TriggerFilterObjectWithRefs> l1TriggeredTaus;
+       event.getByToken(L1TauDesc[inp_idx].input_token, l1TriggeredTaus);
+
+       l1t::TauVectorRef l1Taus;
+       l1TriggeredTaus->getObjects(trigger::TriggerL1Tau, l1Taus);
+       TauCollectionMap.insert(std::pair<size_t, l1t::TauVectorRef>(inp_idx, l1Taus));
+
+       for(auto& i : l1Taus){
+
+         if(std::find(allTaus.begin(), allTaus.end(), i)!= allTaus.end()) continue;
+         allTaus.push_back(i);
+       }
+
+     }
+
+     edm::Handle<EcalRecHitCollection> ebHandle;
+     edm::Handle<EcalRecHitCollection> eeHandle;
+     for (std::vector<edm::EDGetTokenT<EcalRecHitCollection> >::const_iterator i = ecal_tokens.begin(); i != ecal_tokens.end(); i++) {
+       edm::Handle<EcalRecHitCollection> ec_tmp;
+       event.getByToken(*i, ec_tmp);
+       if (ec_tmp->empty())
+         continue;
+       if ((ec_tmp->begin()->detid()).subdetId() == EcalBarrel) {
+         ebHandle = ec_tmp;
+       }
+       else if ((ec_tmp->begin()->detid()).subdetId() == EcalEndcap) {
+         eeHandle = ec_tmp;
+       }
+     }
+
+     std::vector<edm::EDGetTokenT<EcalRecHitCollection> >::const_iterator i;
+     for (i = ecal_tokens.begin(); i != ecal_tokens.end(); i++) {
+       edm::Handle<EcalRecHitCollection> ec;
+       event.getByToken(*i, ec);
+     }
+     edm::Handle<HBHERecHitCollection> hbhe;
+     event.getByToken(hbhe_token, hbhe);
+     edm::Handle<HORecHitCollection> ho;
+     event.getByToken(ho_token, ho);
+     edm::Handle<PixelTrackHeterogeneous> pataTracks;
+     event.getByToken(pataTracks_token, pataTracks);
+     edm::Handle<ZVertexHeterogeneous> pataVertices;
+     event.getByToken(pataVertices_token, pataVertices);
+     edm::Handle<reco::BeamSpot> bsHandle;
+     event.getByToken(BeamSpot_token, bsHandle);
+     edm::ESHandle<MagneticField> fieldESH;
+     eventsetup.get<IdealMagneticFieldRecord>().get(fieldESH);
+     edm::ESHandle<CaloGeometry> Geometry = eventsetup.getHandle(Geometry_token);
+     caloRecHitCollections caloRecHits;
+     caloRecHits.hbhe= &*hbhe;
+     caloRecHits.ho= &*ho;
+     caloRecHits.eb= &*ebHandle;
+     caloRecHits.ee= &*eeHandle;
+     caloRecHits.Geometry = &*Geometry;
+     int evt_id = event.id().event();
+
+     int nTaus = allTaus.size();
+     tensorflow::Tensor cellGridMatrix(tensorflow::DT_FLOAT, { nTaus, nCellEta, nCellPhi, nVars });
+     for (int tau_idx = 0; tau_idx<nTaus ; tau_idx++){
+       for (int eta_idx = 0; eta_idx<nCellEta ; eta_idx++){
+         for (int phi_idx = 0; phi_idx<nCellPhi ; phi_idx++){
+           for (int var_idx = 0; var_idx<nVars ; var_idx++){
+             cellGridMatrix.tensor<float, 4>()(tau_idx,phi_idx, eta_idx,var_idx) = 0.;
+           }
+         }
+       }
+     }
```

Comment:
better to avoid nested fors. E.g.
```c++
const int n_inputs = nTaus * nCellEta * nCellPhi * nVars;
for (int input_idx = 0; input_idx < n_inputs; ++input_idx)
    cellGridMatrix.flat<float>()(input_idx) = 0;

```